tchannel-core/src/main/java/com/uber/tchannel/handlers/RequestRouter.java

```diff
@@ -87,37 +89,27 @@ protected void channelRead0(final ChannelHandlerContext ctx, final Request reque
 
         // There is nothing to do if the connection is already destroyed.
         if (!ctx.channel().isActive()) {
-            logger.warn("drop request when channel is inActive");
             request.release();
+            logger.warn("drop request when channel is inActive");
             return;
         }
 
-        final ArgScheme argScheme = ArgScheme.toScheme(
-                request.getTransportHeaders().get(TransportHeaders.ARG_SCHEME_KEY)
-        );
-
-        if (argScheme == null) {
-            sendError(ErrorType.BadRequest,
-                "Expected incoming call to have \"as\" header set",
-                request, ctx);
+        if (request.getArgScheme() == null) {
+            sendError(ErrorType.BadRequest, "Expected incoming call to have \"as\" header set", request, ctx);
             return;
         }
 
         final String service = request.getService();
         if (service == null || service.isEmpty()) {
-            sendError(ErrorType.BadRequest,
-                "Expected incoming call to have serviceName",
-                request, ctx);
+            sendError(ErrorType.BadRequest, "Expected incoming call to have serviceName", request, ctx);
             return;
         }
 
         // Get the endpoint. The assumption over here is that endpoints are
         // always going to to utf-8 encoded.
-        String endpoint = request.getArg1().toString(CharsetUtil.UTF_8);
+        String endpoint = request.getEndpoint();
         if (endpoint == null || endpoint.isEmpty()) {
```

Comment:
why both checks - null & isEmpty? What is the difference between these? There should be precondition on constructing the request or its setters

Reply:
yes, there should be - but given the lack of nullability contracts in the code it's ok to leave the check in place.
checking for `null` before invoking a method on the object provides protection from `NullPointerException` in the `RequestRouter` code.

Reply:
I understand the NPE check, what is the difference with isEmpty then? I get you point that the code as is should check it.

Reply:
well, `.isEmpty()` checks if the string is empty, i.e. its length is `0`.